hdf5: added +fpic variant

aprokop opened the pull request.

adamjstewart commented:
So the problem here is that not all compilers use `-fpic` to produce position independent code. The option should probably be called `+pic` or `+shared`, since the `-f` is just a flag specific to some compilers. The proper way to do this would be to create a `pic_flag` property for each compiler, just like we did for things like `openmp_flag` and `cxx11_flag`. See the files in `lib/spack/spack/compilers` for examples.

alalazo commented:
This PR doesn't make much sense to me at a first glance. An installation of `zlib` produces the following tree:
```console
/home/mculpo/wdir/spack/opt/spack/linux-ubuntu14-x86_64/gcc-6.2.0/zlib-1.2.8-hahiy3jbdoogfaopwdegikn2dyfkvfjs/
├── include
│   ├── zconf.h
│   └── zlib.h
├── lib
│   ├── libz.a
│   ├── libz.so -> libz.so.1.2.8
│   ├── libz.so.1 -> libz.so.1.2.8
│   ├── libz.so.1.2.8
│   └── pkgconfig
│       └── zlib.pc
└── share
    └── man
        └── man3
            └── zlib.3
```
It installs both shared and static libraries. For shared libraries `pic` is implied, and I don't see the use case for linking static libraries compiled with `pic` support into a shared library upstream. Am I missing something? What is your use case?

aprokop commented:
@adamjstewart Can you point me to a package example satisfying all the requirements?

aprokop commented:
@alalazo An upstream project that I contribute to uses Trilinos with the following options:
```cmake
   -D TPL_ENABLE_Netcdf=ON
        -D TPL_Netcdf_LIBRARIES="libnetcdf.a;libpnetcdf.a;libhdf5_hl.a;libhdf5.a;libz.a"
```
Originally, it used static build for compilation, I'm trying to switch it to a shared build. I did not want to modify the configure too much, and went with fpic option instead. 

Overall, I'm not too familiar with using static libraries and fpic. I would assume it makes sense if one want to use them in shard builds. Are there any drawback to adding such an option?

adamjstewart commented:
@aprokop For the compilers I use:

Compiler | Flag(s) | Notes
--- | --- | ---
GCC | -fpic / -fPIC | -fpic and -fPIC are not equivalent
Intel | -fpic / -fPIC | -fPIC is a synonym for -fpic
PGI | -fpic / -fPIC / -kpic / -kPIC | all are synonyms for -fpic
NAG | -pic / -PIC | -pic and -PIC are not equivalent

As you can see, the NAG Fortran compiler uses a different flag. And since PGI provides `-kpic` "for compatibility with other compilers", it's reasonably safe to assume that other compilers are different.

Take a look at #882 to see what went into adding the `openmp_flag`. In packages like zlib and hdf5, you would access this flag by using `self.compiler.pic_flag`.

alalazo commented:
> Originally, it used static build for compilation, I'm trying to switch it to a shared build.

Did you try to link the shared libraries instead of static ones? Are you having problems with that?

aprokop commented:
> Did you try to link the shared libraries instead of static ones? Are you having problems with that?

At some point I tried but was having trouble with that. I don't recall exactly what happened, so I'm going to try again. But independent of whether that succeeds, does this PR make sense? Or is there an (implicit) assumption for spack that one should not mix shared and static builds?

adamjstewart commented:
We do already use `-fpic` in a lot of other packages already:

adios, arpack, cantera, hdf, metis, mumps, netlib-scalapack, openblas, otf2, parallel-netcdf, parmgridgen, scorep, scotch, suite-sparse, sundials, superlu-dist, and zoltan

alalazo commented:
@aprokop 
> But independent of whether that succeeds, does this PR make sense?

Not to me, but to be fair I have to say this is just my opinion.

> Or is there an (implicit) assumption for spack that one should not mix shared and static builds?

There's no such assumption. We have currently packages with `fpic` variant, but those should be mostly packages that produce only static builds.

@adamjstewart 
>We do already use -fpic in a lot of other packages already:
>
>adios, arpack, cantera, hdf, metis, mumps, netlib-scalapack, openblas, otf2, parallel-netcdf, >parmgridgen, scorep, scotch, suite-sparse, sundials, superlu-dist, and zoltan

Take `arpack` as an example:
```console
$ tree /home/mculpo/wdir/spack/opt/spack/linux-ubuntu14-x86_64/gcc-6.2.0/arpack-96-dgxvchmipvyam47wnerfwpjgki3rpsev
/home/mculpo/wdir/spack/opt/spack/linux-ubuntu14-x86_64/gcc-6.2.0/arpack-96-dgxvchmipvyam47wnerfwpjgki3rpsev
└── lib
    └── libarpack.a
```
In this case you need `pic` code to link into a shared library. For `openblas` for instance I don't understand why it is needed as a shared library is also built. Maybe @davydden has some clue?

adamjstewart commented:
Regardless, we should probably have a `pic_flag`. I'll open a separate issue.

aprokop commented:
@adamjstewart So, should I wait until `pic_flag` is added and then redo the PR?

adamjstewart commented:
You're welcome to add `pic_flag` yourself. Regardless, this PR as is won't work for me because it will crash when I compile HDF 5 with NAG.

davydden commented:
> For openblas for instance I don't understand why it is needed as a shared library is also built. Maybe @davydden has some clues?

as far as i can tell, `+fpic` variant in `openblas` makes differences only when it is `~shared`. So in other words if someone builds static openblas, then there is an option to have PIC.

davydden commented:
@aprokop if you have a look at compiler classes, it is actually straight forward to add `pic_flag`, just look at `/lib/spack/spack/compilers/clang.py` and 
```
    @property
    def openmp_flag(self):
        if self.is_apple:
            tty.die("Clang from Apple does not support Openmp yet.")
        else:
            return "-fopenmp"
```
as an example.

adamjstewart commented:
@davydden I pointed him to #882. The hard part will be handling mixed compilers, as I mentioned in #508.

citibeth commented:
See here for discussion on shared vs. static, and what variants to use for them:

https://groups.google.com/forum/#!topic/spack/gNAp1CmhxF0

In the past, we've added a '+fpic' variant to packages, to allow the building of position independent code for static libraries.  This allows them to be linked with shared libraries.  Since shard libraries are so pervasive, I believe the defaults should be to build shared libraries.  If you say `+static`, it should build a static library too, with PIC code by default.  if you want a static library with non-PIC code, you should say `+static~fpic`.

There is inconsistency about:
  a) Are static libs build by default?
  b) Are shared libs still built if you say `+static`?

I'm happy to renamed `+fpic` to `+pic`.  But that should be done on a Spack-wide basis, to maintain consistency.  I think it should be a separate PR from this one.  All this is orthogonal to the very real benefit in #2375 that we will get from having PIC flags added in a compiler-independent way.

citibeth commented:
> Overall, I'm not too familiar with using static libraries and fpic. I would assume it makes sense if one want to use them in shard builds. Are there any drawback to adding such an option?

In theory, PIC code runs a little slower than non-PIC code.  In practice, there is little discernible difference.  And the PIC code is much more flexible, allowing it to be linked by shared libraries.

becker33 commented on var/spack/repos/builtin/packages/hdf5/package.py:
```diff
@@ -114,6 +116,10 @@ def configure_args(self):
             if spec.satisfies('@:1.8.16'):
                 extra_args.append('--enable-fortran2003')
+        if '+fpic' in spec:
+            extra_args.append('CFLAGS=-fPIC')
```
Can you update this to use the new compiler.pic_flag feature from #2375?

citibeth commented:
Can you also change it to '+pic'?  This conforms with our new standard.

On Fri, Dec 2, 2016 at 2:06 PM, Andrey Prokopenko <notifications@github.com>
wrote:

> *@aprokop* commented on this pull request.
> ------------------------------
>
> In var/spack/repos/builtin/packages/hdf5/package.py:
>
> > @@ -114,6 +116,10 @@ def configure_args(self):
>              # '--enable-fortran2003' no longer exists as of version 1.10.0
>              if spec.satisfies('@:1.8.16'):
>                  extra_args.append('--enable-fortran2003')
> +        if '+fpic' in spec:
> +            extra_args.append('CFLAGS=-fPIC')
>
> Yes, I will redo this.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/LLNL/spack/pull/2373>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB1cdxuT4Hp4oPXXgfLw7Nw7Um1_xc2fks5rEGwygaJpZM4K4lSy>
> .
>

alalazo commented:
@aprokop 
> At some point I tried but was having trouble with that. I don't recall exactly what happened, so I'm going to try again. 

I am curious if you have an update on this :smile:

aprokop commented:
@alalazo I did try shared libraries. The problem I encountered (and that has not been fixed in the recent netcdf update) is that I need netcdf with pnetcdf option. But `parallel-netcdf` only provides a static library. Moreover, netcdf does not find that static version as it tests `-lpnetcdf` without using mpi wrapper which results in undefined references to MPI calls. So, I could not go fully shared libraries route. It may be possible to go hybrid, and use some libraries static (like netcdf) and some shared (like hdf5) but I have not tried that yet.

aprokop commented:
Updated PR.

@citibeth I changed the flag to 'pic'

@becker33 I updated PR to use `pic_flag`.

One question I still not sure about is the interaction between 'pic' and 'shared'. It was suggested to me at some point that when both shared and static libraries are enabled the 'pic' variant should be a no-op, and that it should only be applied when only static libraries are enabled. This PR does not follow that, but I can change it if necessary. From my perspective, I would just enable it unconditionally even when shared libraries are built.

citibeth commented:
> One question I still not sure about is the interaction between 'pic' and
> 'shared'. It was suggested to me at some point that when both shared and
> static libraries are enabled the 'pic' variant should be a no-op, and that
> it should only be applied when only static libraries are enabled. This PR
> does not follow that, but I can change it if necessary. From my
> perspective, I would just enable it unconditionally even when shared
> libraries are built.
>

When you build a shared library, `-fPIC` is added to the command line
(otherwise the shared library won't link).  Therefore, the `pic` variant is
a no-op when `+shared` is enabled.  If you like, you an throw an error
(sine no-op variants change the hash on what is otherwise the same package).

aprokop commented:
@citibeth 

> When you build a shared library, `-fPIC` is added to the command line (otherwise the shared library won't link). Therefore, the `pic` variant is a no-op when `+shared` is enabled.

Sorry, I meant what do we do for static builds in situation when both static and shared builds are enabled? We could build static library with or without a pic flag.

alalazo commented:
>Sorry, I meant what do we do for static builds in situation when both static and shared builds are >enabled? We could build static library with or without a pic flag.

Just for the record, my preferences in order:

1. We don't add `pic` variant for cases where a shared library is built (and of course we fix `netcdf` and `pnetcdf` for your use case). We revert #2470 and we look into cases like `openblas` to see if we can remove `pic`.
2. If we want to add `pic` variants for libraries that already provide a shared version, we should always activate it in the compiler wrapper.
3. We add `pic` variants to every package we have that provides a library (please don't)

aprokop commented:
@alalazo Thank you for summarizing the possible approaches. I have no strong opinion here, and would like to hear other people's thoughts.

> We don't add pic variant for cases where a shared library is built

Did you mean "may be built" here?

becker33 commented:
We already have open issues about policies related to the +shared and +static variants. Feel free to include discussion of their interaction with the +pic/+fpic variants there, but that should not hold up integration of these changes for now.
